mempool/reactor_test.go

```diff
@@ -333,6 +334,43 @@ func TestReactorTxSendersMultiNode(t *testing.T) {
 	require.Zero(t, len(firstReactor.txSenders))
 }
 
+func TestMempoolFIFOWithParallelCheckTx(t *testing.T) {
+	if os.Getenv("CI") != "" {
+		// We expect this test to fail, but do not want to block merges because of it.
+		t.Skip("Skipping testing in CI environment")
+	}
+
+	config := cfg.TestConfig()
+	reactors, _ := makeAndConnectReactors(config, 4)
+	defer func() {
+		for _, r := range reactors {
+			if err := r.Stop(); err != nil {
+				assert.NoError(t, err)
+			}
+		}
+	}()
+	for _, r := range reactors {
+		for _, peer := range r.Switch.Peers().List() {
+			peer.Set(types.PeerStateKey, peerState{1})
+		}
+	}
+
+	// Deliver the same sequence of transactions from multiple sources, in parallel.
+	txs := newUniqueTxs(20)
+	mp := reactors[0].mempool
+	for i := 0; i < 3; i++ {
+		go func() {
+			for _, tx := range txs {
+				_, err := mp.CheckTx(tx)
+				assert.Nil(t, err)
```

Comment:
I'm getting an error "tx already exists in cache", which is expected but it's not the goal of this test.

Reply:
Thanks @hvanz
Fixe the code to not break on this error, just to output it and confirmed that it fails because of the inversion only.